core/modules/wsched/ChunkTaskList.h

```diff
@@ -0,0 +1,160 @@
+        static bool compareFunc(wbase::Task::Ptr const& x, wbase::Task::Ptr const& y) {
+            if(!x || !y) { return false; }
+            // compare scanInfo (slower scans first)
+            int siComp = x->getScanInfo().compareTables(y->getScanInfo());
+            return siComp < 0;
+        };
+        void push(wbase::Task::Ptr const& task);
+        wbase::Task::Ptr pop();
+        wbase::Task::Ptr top() {
+            if (_tasks.empty()) return nullptr;
+            return _tasks.front();
+        }
+        bool empty() const { return _tasks.empty(); }
+        size_t size() const { return _tasks.size(); }
+        void heapify() {
+            std::make_heap(_tasks.begin(), _tasks.end(), compareFunc);
+        }
+
+        std::vector<wbase::Task::Ptr> _tasks;
+    };
+
+private:
+    int _chunkId;          ///< Chunk Id for all Tasks in this instance.
+    bool _active{false};   ///< True when this is the active chunk.
+    bool _resourceStarved; ///< True when advancement is prevented by lack of memory.
+    wbase::Task::Ptr              _readyTask{nullptr}; ///< Task that is ready to run with memory reserved.
+    SlowTableHeap                 _activeTasks;        ///< All Tasks must be put on this before they can run.
+    std::vector<wbase::Task::Ptr> _pendingTasks;       ///< Task that should not be run until later.
+    std::set<wbase::Task*>        _inFlightTasks;      ///< Set of Tasks that this chunk has in flight.
+
+    memman::MemMan::Ptr _memMan;
+};
+
+
+/// This class queues Tasks by their chunkId and tables rating and names.
+/// New Tasks are queued with other Tasks with the same chunkId and then by shared
+/// scan tables used.
+/// - Tasks are provided starting with the _activeChunk, which remains the
+///   _activeChunk until all of its Tasks are completed. At which time, the
+///   _activeChunk advances to the chunk with the next higher chunkId. While
+///   a chunk is the _activeChunk, all new Tasks for that chunk are put in
+///   in a pending list so that the active chunk does not get stalled.
+/// - While all the Tasks on the _active chunk have been started, but not completed,
+///   Tasks can be taken from chunks after the _activeChunk as long as resources are
+///   available.
+/// Like the other schedulers, ready() is the core of this class as it determines
+/// if a Task is ready to run and which Task will be provided by getTask().
+class ChunkTaskList : public ChunkTaskCollection {
+public:
+    using Ptr = std::shared_ptr<ChunkTaskList>;
+
+    /// This must be std::map to maintain valid iterators.
+    /// Only erase() will invalidate and iterator with std::map.
+    using ChunkMap = std::map<int, ChunkTasks::Ptr>;
+
+    enum {READY, NOT_READY, NO_RESOURCES};
+
+    ChunkTaskList(memman::MemMan::Ptr const& memMan) : _memMan{memMan} {}
+    ChunkTaskList(ChunkTaskList const&) = delete;
+    ChunkTaskList& operator=(ChunkTaskList const&) = delete;
```

Comment:
The comments would be redundant for most of these do exactly what the function name says they will do. In my opinion, such comments just make the header harder to read. Specifics to the implementation are more appropriate in the cc files. 

I did add descriptions to some of these in ChunkTaskCollection.